src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c

```diff
@@ -839,11 +839,28 @@ int32_t GlobalizationNative_LastIndexOf(
 
     result = usearch_last(pSearch, &err);
 
-    // if the search was successful,
-    // we'll try to get the matched string length.
-    if (result != USEARCH_DONE && pMatchedLength != NULL)
+    // if the search was successful, we'll try to get the matched string length.
+    if (result != USEARCH_DONE)
     {
-        *pMatchedLength = usearch_getMatchedLength(pSearch);
+        if (pMatchedLength != NULL)
+        {
+            *pMatchedLength = usearch_getMatchedLength(pSearch);
+        }
+
+        // In case the search result pointing at last character of the source string, we need to check if the target string
```

Comment:
Does this "special-case pointing to the last char" logic also work when the string ends with surrogates, glyphs that consist of a single base character followed by 1 or more accent / modifier characters, etc., before the final zero-weight terminator?

Reply:
at that time the match length shouldn't be zero. but I'll be happy to try it if you provide example of that.

Reply:
I have tried 

```C#
    Console.WriteLine($"***************** {"A\u0303".LastIndexOf("\u200D")}"); // A +  ̃ = Ã
    Console.WriteLine($"***************** {"A\u0303\u200D".LastIndexOf("\u200D")}"); // A +  ̃ = Ã
```

and it prints 

```
  ***************** 2
  ***************** 3
```

The result is consistent between ICU and NLS. 

@GrabYourPitchforks is that what you meant to try?

Reply:
The `"A\u0303"` sample is great. We should consider another test case: `"😁".LastIndexOf(<zero-weight-string>)`. Since the string consists of an emoji (surrogate pair), it'd be useful to know if the iterator returns "length - 1" or "length - 2" to account for the surrogate pair. If it's the latter, would our logic break?

Reply:
Good test case to try. I have added a fix for the surrogate cases and included more tests to cover different cases mentioned in this comment. Thanks!